interface.py

```diff
@@ -0,0 +1,193 @@
+import argparse
+import importlib
+import os
+import sys
+from datetime import datetime, timedelta
+
+import pandas as pd
+import pytz
+from sqlalchemy import create_engine
+
+raw_data_tables = [
+    'applications_foreground',
+    'battery',
+    'calls',
+    'light',
+    'locations',
+    'messages',
+    'screen'
+]
+
+
+def read_data_func(datastore, container):
+    if datastore == 'sql':
+        return lambda table_name: pd.read_sql('select * from {}'.format(table_name), container)
+    elif datastore == 'csv':
+        return lambda table_name: pd.read_csv(os.path.join(container, '{}.csv'.format(table_name)), sep='\t')
+    else:
+        raise TypeError('Unrecognized type: ' + str(datastore))
+
+
+def preprocess_aware(table, timezone_table):
+    # Restructure the timezone table to have a beginning (inclusive) and ending (exclusive) timestamp.
+    # The ending timestamp is the next timestamp found for that device. If there is no next timestamp, use the latest
+    # timestamp from the table to be processed + 1.
+    max_timestamp = table['timestamp'].max() + 1
+    timezone_table['end_timestamp'] = timezone_table \
+        .sort_values(['device_id', 'timestamp']) \
+        .groupby('device_id')['timestamp'] \
+        .shift(-1) \
+        .rename(columns={'timestamp': 'start_timestamp'}) \
+        .fillna(max_timestamp) \
+        .drop('_id')
+
+    # Assumes we have an initial timezone at recruitment time, in which case all values in the sensor table should
+    # occur between the earliest timezone timestamp and the latest sensor table timestamp (which is also the latest
+    # timezone stamp).
+    table = table.merge(timezone_table, on='device_id')
+    table = table[(table['timestamp'] >= table['start_timestamp']) & (table['timestamp'] < table['end_timestamp'])]
+    table['time_offset'] = table.apply(lambda r: pytz
+                                       .timezone(r['timezone'])
+                                       .utcoffset(datetime.fromtimestamp(r['timestamp'] // 1000)), axis='columns')
```

Comment:
This assumes the offset comes in seconds, and the timestamp in ms, correct?

Reply:
Yes, which I think is a mistake on my part. AWARE timestamps are in milliseconds, so I intended to convert the offset back to milliseconds (Python only gives it to you in seconds). But is PDK offset in seconds or milliseconds? If it's seconds, maybe we should standardize to seconds instead.